src/coreComponents/dataRepository/BufferOpsDevice.cpp

```diff
@@ -187,22 +199,22 @@ UnpackDataByIndexDevice ( buffer_unit_type const * & buffer,
                           parallelDeviceEvents & events )
 {
   localIndex numIndices = indices.size();
-  buffer_unit_type const * devBuffer = buffer;
-  localIndex unitSize = var.size() / var.size( 0 ) * sizeof(T);
+  localIndex sliceSize = var.size() / var.size( 0 );
+  localIndex unpackSize = numIndices * sliceSize * sizeof( T );
+  T const * devBuffer = reinterpret_cast< T const * >( buffer );
   parallelDeviceStream stream;
-  events.emplace_back( forAll< parallelDeviceAsyncPolicy<> >( stream, numIndices, [=] GEOSX_DEVICE ( localIndex const i )
+  events.emplace_back( forAll< parallelDeviceAsyncPolicy<> >( stream, numIndices, [=] GEOSX_DEVICE ( localIndex const ii )
   {
-    buffer_unit_type const * threadBuffer = devBuffer + i * unitSize;
-    LvArray::forValuesInSlice( var[ indices[ i ] ], [&threadBuffer] GEOSX_DEVICE ( T & value )
+    T const * threadBuffer = &devBuffer[ ii * sliceSize ];
+    LvArray::forValuesInSlice( var[ indices[ ii ] ], [&threadBuffer] GEOSX_DEVICE ( T & value )
```

Comment:
I was thinking about it later and you can only get rid of `forValuesInSlice` if the slice is contiguous. If we don't care the order that it's packed in, as long as it's unpacked in a consistent manner then we can probably do much better but it would take some thinking as to how best to pack things and it would likely involve a single thread packing parts of multiple objects in some cases.